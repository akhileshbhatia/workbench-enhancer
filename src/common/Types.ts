export type ChromeStorageQueryData = {
  data: string
};

export type TimeDetailsMap = Map<number, ChromeStorageQueryData>;

export type QueryDataMap = Map<string, TimeDetailsMap>;
