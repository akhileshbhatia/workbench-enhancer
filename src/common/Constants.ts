export const extensionStateKey = 'extension_states';
export const appId = 'wb-enhancer-app';
export const mainBlockId = 'mainBlock';
