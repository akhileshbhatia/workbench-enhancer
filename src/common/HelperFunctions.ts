import { QueryDataMap, ChromeStorageQueryData, TimeDetailsMap } from './Types';

export function getDataFromChromeStorage(key: string): Promise<Record<string, unknown>> {
  return new Promise((resolve, reject) => {
    try {
      chrome.storage.local.get(key, (data) => {
        key === null ? resolve(data) : resolve(data[key]);
      });
    } catch (err) {
      reject(err);
    }
  });
}

export function clearStorage(): Promise<void> {
  return new Promise((resolve, reject) => {
    try {
      chrome.storage.local.clear(() => resolve());
    } catch (err) {
      reject(err);
    }
  });
}

export function setDataToChromeStorage(key: string, data: any): Promise<void> {
  return new Promise((resolve, reject) => {
    try {
      chrome.storage.local.set({ [key]: data }, () => {
        if (chrome.runtime.lastError) {
          reject('Error in setting data to chrome storage');
        }
        resolve();
      });
    } catch (err) {
      reject(err);
    }
  });
}

export function serializeMap<keyType, valueType>(map: Map<keyType, valueType>): string {
  return JSON.stringify(Array.from(map.entries()));
}

export function deserializeToMap<keyType, valueType>(data: string): Map<keyType, valueType> {
  return new Map(JSON.parse(data));
}

export function deserializeData(data: string): { output: QueryDataMap } {
  const finalMap = new Map();
  if (data) {
    const dateMap = deserializeToMap<string, string>(data);
    for (const [date, info] of dateMap.entries()) {
      const newValue = deserializeToMap<number, ChromeStorageQueryData>(info);
      finalMap.set(date, newValue);
    }
  }
  return { output: finalMap };
}

export function getHoursAndMinsFromTimestamp(timestamp: number): string {
  const date = new Date(timestamp * 1000);
  // Pad 0 to hours and mins in case they are single digit
  const hours = String(date.getHours()).padStart(2, '0');
  const mins = String(date.getMinutes()).padStart(2, '0');
  return `${hours}:${mins}`;
}

export function getFormattedDateAndTimestamp(): {
  formattedDate: string,
  timestamp: number
} {
  const todaysDate = new Date();
  const year = new Intl.DateTimeFormat('en', { year: 'numeric' }).format(todaysDate);
  const month = new Intl.DateTimeFormat('en', { month: 'short' }).format(todaysDate);
  const day = new Intl.DateTimeFormat('en', { day: '2-digit' }).format(todaysDate);
  return {
    formattedDate: `${day} ${month} ${year}`,
    timestamp: Math.round(<number><unknown>todaysDate / 1000)
  };
}

export function updateTextArea(value: string): void {
  document.querySelector('textarea').value = value;
}

export async function formatDataAndAddtoStorage(
  dataArray: [string, TimeDetailsMap][],
  currentPathName: string
): Promise<void> {
  const dataToStore = new Map<string, string>();
  for (const [date, details] of dataArray) { // Serialize data again before storing
    dataToStore.set(date, serializeMap<number, Record<string, unknown>>(details));
  }
  await setDataToChromeStorage(currentPathName, serializeMap<string, string>(dataToStore));
}
