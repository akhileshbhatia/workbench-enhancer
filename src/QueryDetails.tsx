import React, { ReactElement } from 'react';
import './app.scss';
import { getHoursAndMinsFromTimestamp, updateTextArea } from './common/HelperFunctions';
import DeleteIcon from '@material-ui/icons/Delete';
import StarIcon from '@material-ui/icons/StarBorder';
import { QueryData } from './QueryData';
import { ChromeStorageQueryData } from './common/Types';

export type QueryDetailsProps = {
  timestamp: number,
  details: ChromeStorageQueryData,
  handleDelete: (timestamp: number) => Promise<void>
};

export function QueryDetails(props: QueryDetailsProps): ReactElement {
  const { timestamp, details } = props;

  return (
    <div className='row'>
      <div className='column-flex-2'>
        <b>{getHoursAndMinsFromTimestamp(timestamp)}</b>
      </div>
      <div className='column-flex-6' data-testid="querydata-component" onClick={() => updateTextArea(details.data)}>
        <QueryData {...details} />
      </div>
      <div className='column-flex-1' data-testid="delete-icon" onClick={() => props.handleDelete(timestamp)}>
        <DeleteIcon />
      </div>
      <div className='column-flex-1' data-testid="bookmark-icon">
        <StarIcon />
      </div>
    </div>
  );
}
