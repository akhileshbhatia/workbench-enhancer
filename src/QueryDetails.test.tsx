import React from 'react';
import { RenderResult, render, fireEvent, getByTestId } from '@testing-library/react';
import { QueryDetailsProps, QueryDetails } from './QueryDetails';
import { updateTextArea, getHoursAndMinsFromTimestamp } from './common/HelperFunctions';

jest.mock('./common/HelperFunctions');

describe('QueryDetails', () => {
  let component: RenderResult;
  let handleDelete;

  const init = (): void => {
    const props: QueryDetailsProps = {
      timestamp: 1411070714,
      details: { data: 'test-data' },
      handleDelete
    };
    component = render(<QueryDetails {...props} />);
  };

  beforeEach(() => {
    jest.clearAllMocks();
    handleDelete = jest.fn();
    (getHoursAndMinsFromTimestamp as jest.Mock).mockReturnValue('10:45');
    init();
  });

  it('renders the component correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('calls the handleDelete function', () => {
    fireEvent.click(getByTestId(component.container, 'delete-icon'));
    expect(handleDelete).toHaveBeenCalled();
  });

  it('calls the updateTextArea function', () => {
    fireEvent.click(getByTestId(component.container, 'querydata-component'));
    expect((updateTextArea as jest.Mock)).toHaveBeenCalledTimes(1);
    expect((updateTextArea as jest.Mock)).toHaveBeenCalledWith('test-data');
  });

});
