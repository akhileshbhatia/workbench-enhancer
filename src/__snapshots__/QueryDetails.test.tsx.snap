// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryDetails renders the component correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="row"
      >
        <div
          class="column-flex-2"
        >
          <b>
            10:45
          </b>
        </div>
        <div
          class="column-flex-6"
          data-testid="querydata-component"
        >
          <div>
            test-data
          </div>
        </div>
        <div
          class="column-flex-1"
          data-testid="delete-icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
        </div>
        <div
          class="column-flex-1"
          data-testid="bookmark-icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
            />
          </svg>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="row"
    >
      <div
        class="column-flex-2"
      >
        <b>
          10:45
        </b>
      </div>
      <div
        class="column-flex-6"
        data-testid="querydata-component"
      >
        <div>
          test-data
        </div>
      </div>
      <div
        class="column-flex-1"
        data-testid="delete-icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </div>
      <div
        class="column-flex-1"
        data-testid="bookmark-icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
          />
        </svg>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
